queenbee/schema/function.py

```diff
@@ -22,92 +132,123 @@ class Function(BaseModel):
         ' this function.'
     )
 
-    inputs: Arguments = Field(
-        None,
+    operator: str = Field(
+        ...,
+        description='Function operator name.'
+    )
+    
+    inputs: FunctionInputs = Field(
+        FunctionInputs(),
         description=u'Input arguments for this function.'
     )
 
     command: str = Field(
-        ...,
+        '...',
         description=u'Full shell command for this function. Each function accepts only '
         'one command. The command will be executed as a shell command in operator. '
         'For running several commands after each other use && between the commands '
         'or pipe data from one to another using |'
     )
 
-    operator: str = Field(
-        ...,
-        description='Function operator name.'
+    outputs: FunctionOutputs = Field(
+        FunctionOutputs(),
+        description='List of output arguments.'
     )
 
-    # TODO: check for referenced variables
-    env: Dict[str, str] = Field(
-        None,
-        description='A dictionary of key:values for environmental variables.'
-    )
+    @classmethod
+    def validate_referenced_values(cls, input_names: List[str], variables: List):
+        """Validate referenced values"""
+
+        for ref in variables:
+            add_info = ''
+            ref = ref.replace('{{', '').replace('}}', '').strip()
+            add_info = qbvar._validate_inputs_outputs_var_format(ref)
```

Comment:
We should reuse the methods from `parser` modules or other modules that we have. Using find and replace here and there will make it very hard to debug it in the long term.

Reply:
I'm on the fence about this one as it turns out that quite a few of those functions were not re-used in any other schema. ie: there were some parser functions for workflows, some for functions and others for DAGs. As such it felt like it made sense to just stick them in the matching schema classes.

That being said this might change if we move from using `{{...}}` notation to reference objects as we discussed a few days ago. 

Mind waiting a few commits to see what it looks like in a few days?